store.go

```diff
@@ -2962,6 +3019,13 @@ func (s *store) Image(id string) (*Image, error) {
 	if done, err := s.readAllImageStores(func(store roImageStore) (bool, error) {
 		image, err := store.Get(id)
 		if err == nil {
+			if !store.(*imageStore).lockfile.IsReadWrite() {
```

Comment:
I’d rather prefer `store != s.imageStore` instead of this condition. `lockfile` is shared across all users in the _process_, even across `imageStore` and `store` instances, so this implicitly relies on `getLockfile` only allowing a single lock to be only referenced read-write / read-only consistently. That’s a bit remote dependency. (And, pedantically, the Windows lock implementation always returns `IsReadWrite() == false`.)

---

Actually, this _should_ AFAICS be equivalent to `if false`, because `s.imageStore` is the first store used by `readAllImageStores`. So maybe just explicitly document that aspect of `allImageStores` and `readAllImageStores`, and point out that we rely on this here in a comment.

Reply:
It shouldn't be equivalent to `if false`.  The `lockfile` for a given `imageStore` struct is initialized with its `ro` field set to `false` for a writable store, and `true` for a read-only store, and supposedly that never changes.  We depend on this code path to hide the read-only version of an image that still has a name on it that we intended to remove; one of the new tests exercises that scenario.

Reply:
My reading of the code is
- `s.imageStore` is only ever initialized from `newImageStore` , which gets a read-write lock (and fails with an error if the same lock is already used read-only)
- Elements of `s.roImageStores` are only ever initialized from `newROImageStore`, similarly clearly read-only.
- `store.readAllImageStores` processes them in the order of `store.allImageStores`, i.e. read-write first, read-only afterwards
- So, if the image exists in the (only) writable store, we find it in the first iteration, and immediately exit the loop (by returning `done = true` from the closure).
  - We will never have the opportunity to find a duplicate image in any read-only store, i.e. the newly added code is never reached.
  - Alternatively, even if that code were reached (which cannot be the case), the second `s.imageStore.Get` of the read-write store must report an error, because we have already processed `s.imageStore` as the first item, and didn’t find the image there (and the store is still locked)

Which part is incorrect?

---

… does this hinge on the difference between `Get(id)` and `Get(image.ID)`, in a situation where the `id` parameter is not actually an ID but a name or an ID prefix?

Oh. That was totally non-obvious to me.

- primarily because the documentation of `Store.Image` just names the parameter `id` and doesn’t say anything about names — but yes, we do use that method to look up names.
- secondarily the new “no, we didn’t” comment hints at, but doesn’t _quite say_, that the first lookup was probably by a name and that name now doesn’t apply.

Do I understand the intent correctly now, or is there even more that I’m missing?

Reply:
Yup, will add to the comment that it's a lookup-by-name case.

Reply:
Thanks!

WDYT about 
```suggestion
			if store != s.imageStore
```
here? It seems to me that the code assumes that there is only one read-write store, and that it’s `s.imageStore`, already.

Reply:
It's not the call to a lockfile's `IsReadWrite()` method that we're using in most other places to check for writability... but it avoids a type assertion.  Changing it.